File: src/main/java/lambda/function/EqFunction.java
package lambda.function;

import lambda.Function;
import lambda.core.datamodels.AtomNode;

interface EqFunction extends Function {

    static EqFunction newInstance() {
        return params -> {
            final var first = params.pop();
            while(params.isNotEmpty()) {
                final var node = params.pop();
                if (!node.equals(first)) {
                    return new AtomNode(false);
                }
            }
            return new AtomNode(true);
        };
    }
}

File: src/main/java/lambda/function/EqFunction.kt
package lambda.function

import lambda.core.datamodels.AtomNode
import lambda.core.datamodels.Stack
import lambda.core.datamodels.Node

class EqFunction : Function {
    override fun evaluate(
        params: Stack<Node>
    ): Node {
        val first = params.pop()

        while(params.isNotEmpty()) {
            val node = params.pop()
            if (node != first) {
                return AtomNode(false)
            }
        }
        return AtomNode(true)
    }
}